test/functional/eval/writefile_spec.lua

```diff
@@ -0,0 +1,93 @@
+local helpers = require('test.functional.helpers')(after_each)
+
+local clear = helpers.clear
+local eq = helpers.eq
+local funcs = helpers.funcs
+local exc_exec = helpers.exc_exec
+local read_file = helpers.read_file
+
+local fname = 'Xtest-functional-eval-writefile'
+
+before_each(clear)
+after_each(function() os.remove(fname) end)
+
+describe('writefile()', function()
+  it('writes empty list to a file', function()
+    eq(nil, read_file(fname))
+    eq(0, funcs.writefile({}, fname))
+    eq('', read_file(fname))
+    os.remove(fname)
+    eq(nil, read_file(fname))
+    eq(0, funcs.writefile({}, fname, 'b'))
+    eq('', read_file(fname))
+    os.remove(fname)
+    eq(nil, read_file(fname))
+    eq(0, funcs.writefile({}, fname, 'ab'))
+    eq('', read_file(fname))
+    os.remove(fname)
+    eq(nil, read_file(fname))
+    eq(0, funcs.writefile({}, fname, 'a'))
+    eq('', read_file(fname))
+  end)
+
+  it('writes list with an empty string to a file', function()
+    eq(0, exc_exec(
+       ('call writefile([$XXX_NONEXISTENT_VAR_XXX], "%s", "b")'):format(
+          fname)))
+    eq('', read_file(fname))
+    eq(0, exc_exec(('call writefile([$XXX_NONEXISTENT_VAR_XXX], "%s")'):format(
+        fname)))
+    eq('\n', read_file(fname))
+  end)
+
+  it('appends to a file', function()
+    eq(nil, read_file(fname))
+    eq(0, funcs.writefile({'abc', 'def', 'ghi'}, fname))
+    eq('abc\ndef\nghi\n', read_file(fname))
+    eq(0, funcs.writefile({'jkl'}, fname, 'a'))
+    eq('abc\ndef\nghi\njkl\n', read_file(fname))
+    os.remove(fname)
+    eq(nil, read_file(fname))
+    eq(0, funcs.writefile({'abc', 'def', 'ghi'}, fname, 'b'))
+    eq('abc\ndef\nghi', read_file(fname))
+    eq(0, funcs.writefile({'jkl'}, fname, 'ab'))
+    eq('abc\ndef\nghijkl', read_file(fname))
+  end)
+
+  it('correctly treats NLs', function()
+    eq(0, funcs.writefile({'\na\nb\n'}, fname, 'b'))
+    eq('\0a\0b\0', read_file(fname))
+    eq(0, funcs.writefile({'a\n\n\nb'}, fname, 'b'))
+    eq('a\0\0\0b', read_file(fname))
+  end)
+
+  it('correctly overwrites file', function()
```

Comment:
Suggestion: "correctly" and "successfully" are redundant in test descriptions, it should be assumed that we (almost) always want correct behavior.